peps/pep-0788.rst

```diff
@@ -0,0 +1,686 @@
+  thread local storage.
+
+In the current C API, any non-Python thread (one not created via the
+:mod:`threading` module) is considered to be "daemon", meaning that the interpreter
+won't wait on that thread to finalize. Instead, the interpreter will hang the
+thread when it goes to :term:`attach <attached thread state>` a :term:`thread state`,
+making it unusable past that point. Attaching a thread state can happen at
+any point when invoking Python, such as releasing the GIL in-between bytecode
+instructions, or when a C function exits a :c:macro:`Py_BEGIN_ALLOW_THREADS`
+block. (Note that hanging the thread is relatively new behavior; in prior
+versions, the thread would terminate, but the issue is the same.)
+
+This means that any non-Python thread may be terminated at any point, which
+is severely limiting for users who want to do more than just execute Python
+code in their stream of calls (for example, C++ executing finalizers in
+*addition* to calling Python).
+
+Using :c:func:`Py_IsFinalizing` is insufficient
+***********************************************
+
+The :c:func:`docs <PyGILState_Ensure>`
+currently recommend :c:func:`Py_IsFinalizing` to guard against termination of
+the thread:
+
+    Calling this function from a thread when the runtime is finalizing will
+    terminate the thread, even if the thread was not created by Python. You
+    can use ``Py_IsFinalizing()`` or ``sys.is_finalizing()`` to check if the
+    interpreter is in process of being finalized before calling this function
+    to avoid unwanted termination.
+
+Unfortunately, this isn't correct, because of time-of-call to time-of-use
+issues; the interpreter might not be finalizing during the call to
+:c:func:`Py_IsFinalizing`, but it might start finalizing right after, which
+would cause the attachment of a thread state (typically via
+:c:func:`PyGILState_Ensure`) to hang the thread.
+
+Daemon threads can cause finalization deadlocks
+***********************************************
+
+When acquiring locks, it's extremely important to detach the thread state to
+prevent deadlocks. This is true on both the with-GIL and free-threaded builds.
+When the GIL is enabled, a deadlock can occur pretty easily when acquiring a
+lock if the GIL wasn't released, and lock-ordering deadlocks can still occur
+free-threaded builds if the thread state wasn't detached.
+
+So, all code that needs to work with locks need to detach the thread state.
+In C, this is almost always done via :c:macro:`Py_BEGIN_ALLOW_THREADS` and
+:c:macro:`Py_END_ALLOW_THREADS`, in a code block that looks something like this:
+
+.. code-block:: c
+
+    Py_BEGIN_ALLOW_THREADS
+    acquire_lock();
+    Py_END_ALLOW_THREADS
+
+Again, in a daemon thread, :c:macro:`Py_END_ALLOW_THREADS` will hang the thread
+if the interpreter is finalizing. But, :c:macro:`Py_BEGIN_ALLOW_THREADS` will
+*not* hang the thread; the lock will be acquired, and *then* hung! Once that
```

Comment:
I think the phrasing could be clearer here (specifically, the sentence reads as that the lock hangs, but I don't think that's what you mean). 

I'm also not particularly keen on 'hang' as a verb (throughout), but it might be more common as a term of art. (I would phrase it as 'X will not cause the thread to hang' rather than 'X will not hang the thread'.)